tests/drivers/dma/scatter_gather/src/test_dma_sg.c

```diff
@@ -0,0 +1,155 @@
+ *   -# Set direction memory-to-memory with two block transfers
+ *   -# Start transfer tx -> rx
+ * - Expected Results
+ *   -# Data is transferred correctly from src buffers to dest buffers without
+ *      software intervention.
+ */
+
+#include <kernel.h>
+#include <drivers/dma.h>
+#include <ztest.h>
+
+#define XFERS 2
+#define XFER_SIZE 64
+
+#if CONFIG_NOCACHE_MEMORY
+static const char TX_DATA[] = "The quick brown fox jumps over the lazy dog";
+static __aligned(32) char tx_data[XFER_SIZE] __used
+	__attribute__((__section__(".nocache")));
+static __aligned(32) char rx_data[XFERS][XFER_SIZE] __used
+	__attribute__((__section__(".nocache.dma")));
+#else
+/* this src memory shall be in RAM to support usingas a DMA source pointer.*/
+static const char tx_data[] = "The quick brown fox jumps over the lazy dog";
+static __aligned(32) char rx_data[XFERS][XFER_SIZE] = { { 0 } };
+#endif
+
+#define DMA_DEVICE_NAME CONFIG_DMA_SG_DRV_NAME
+
+K_SEM_DEFINE(xfer_sem, 0, 1);
+
+static struct dma_config dma_cfg = {0};
+static struct dma_block_config dma_block_cfgs[XFERS];
+
+static void dma_sg_callback(const struct device *dma_dev, void *user_data,
+			    uint32_t channel, int status)
+{
+	if (status) {
+		TC_PRINT("callback status %d\n", status);
+	} else {
+		TC_PRINT("giving xfer_sem\n");
+		k_sem_give(&xfer_sem);
+	}
+}
+
+static int test_sg(void)
+{
+	const struct device *dma;
+	static int chan_id;
+
+	TC_PRINT("DMA memory to memory transfer started on %s\n",
+	       DMA_DEVICE_NAME);
+	TC_PRINT("Preparing DMA Controller\n");
+
+#if CONFIG_NOCACHE_MEMORY
+	memset(tx_data, 0, sizeof(tx_data));
+	memcpy(tx_data, TX_DATA, sizeof(TX_DATA));
+#endif
+	memset(rx_data, 0, sizeof(rx_data));
+
+	dma = device_get_binding(DMA_DEVICE_NAME);
+	if (!dma) {
+		TC_PRINT("Cannot get dma controller\n");
+		return TC_FAIL;
+	}
+
+	dma_cfg.channel_direction = MEMORY_TO_MEMORY;
+	dma_cfg.source_data_size = 4U;
+	dma_cfg.dest_data_size = 4U;
+	dma_cfg.source_burst_length = 4U;
+	dma_cfg.dest_burst_length = 4U;
+#ifdef CONFIG_DMAMUX_STM32
+	dma_cfg.user_data = (struct device *)dma;
+#else
+	dma_cfg.user_data = NULL;
+#endif /* CONFIG_DMAMUX_STM32 */
+	dma_cfg.dma_callback = dma_sg_callback;
+	dma_cfg.block_count = XFERS;
+	dma_cfg.head_block = dma_block_cfgs;
+	dma_cfg.complete_callback_en = false; /* per block completion */
+
+#ifdef CONFIG_DMA_MCUX_TEST_SLOT_START
+	dma_cfg.dma_slot = CONFIG_DMA_MCUX_TEST_SLOT_START;
+#endif
+
+	chan_id = dma_request_channel(dma, NULL);
+	if (chan_id < 0) {
+		TC_PRINT("Platform does not support dma request channel,"
+			 " using Kconfig DMA_SG_CHANNEL_NR\n");
+		chan_id = CONFIG_DMA_SG_CHANNEL_NR;
+	}
+
+	memset(dma_block_cfgs, 0, sizeof(dma_block_cfgs));
+	for (int i = 0; i < XFERS; i++) {
+		dma_block_cfgs[i].source_gather_en = 1U;
+		dma_block_cfgs[i].block_size = XFER_SIZE;
+		dma_block_cfgs[i].source_address = (uint32_t)(tx_data);
+		dma_block_cfgs[i].dest_address = (uint32_t)(rx_data[i]);
```

Comment:
Can't these address be 64bit if `CONFIG_DMA_64BIT` is selected ? Btw, why don't we use `uintptr_t` for these address instead of having it conditionally set ? (not related with this particular patch, just a question)